File: USACO_Dec2023/FarmerJohnActuallyFarms.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

bool check(vector<pair<int, pair<ll,ll>>>& plants, int n) {
    for (int i = 1; i < n; i++) {
        if (plants[i-1].second.first <= plants[i].second.first)
            return false;
    }

    return true;
}

void solve() {
    int n; cin >> n;

    vector<pair<int, pair<ll,ll>>> plants(n);
    for (pair<int, pair<ll, ll>>& p : plants) { // getting the heights
        cin >> p.second.first;
    }
    for (pair<int, pair<ll, ll>>& p : plants) { // getting the growthrates
        cin >> p.second.second;
    }
    for (pair<int, pair<ll, ll>>& p : plants) { // getting the order it should be in
        cin >> p.first;
    }

    if (n == 1) { // corner case
        if (plants[0].first == 0)
            cout << 0 << endl;
        else 
            cout << -1 << endl;
        return;
    }

    sort(plants.begin(), plants.end());

    for (int i = 1; i < n; i++) {
        if (plants[i-1].second.second == plants[i].second.second) {
            if (plants[i-1].second.first <= plants[i].second.first) {
                cout << -1 << endl;
                return;
            }
        }

        // if it has a zero growth rate
        if (plants[i-1].second.second == 0) {
            if (plants[i-1].second.first <= plants[i].second.first) {
                cout << -1 << endl;
                return;
            }
        }

        // if it has a negative growth rate
        if (plants[i-1].second.second < 0) {
            if (plants[i-1].second.first <= plants[i].second.first) {
                cout << -1 << endl;
                return;
            }
        }

        // if it has a smaller growth rate than the previous
        if (plants[i-1].second.second < plants[i].second.second && plants[i-1].second.first < plants[i].second.first) {
                cout << -1 << endl;
                return;
        }
    }

    int days = 0;
    while(true) {
        if (check(plants, n)) {
            break;
        }

        for (int i = 0; i < n; i++) {
            plants[i].second.first += plants[i].second.second;
        }

        days++; 
        if (days > 100000) {
            days = -1;
            break;
        };
    }

    cout << days << endl;

}

int main() {
    int t; cin >> t;
    while (t) {
        solve();
        t--;
    }
}

File: USACO_Dec2023/FarmerJohnActuallyFarms.py
def solve():
    

    plants = []
    for _ in range(n):  # getting the order it should be in
        p = int(input())
        plants.append(p)

    if n == 1:  # corner case
        print(0)
        return

    plants.sort()

    # check if the growthrates are in descending order
    for i in range(1, n):
        if plants[i-1][1][1] < plants[i][1][1]:
            print(-1)
            return
        elif plants[i-1][1][1] == plants[i][1][1]:
            if plants[i-1][1][0] <= plants[i][1][0]:
                print(-1)
                return

    days = 0
    while True:
        if check(plants, n):
            break

        for i in range(n):
            plants[i][1][0] += plants[i][1][1]

        days += 1

    print(days)

t = int(input())
for range(t):
    solve()